src/cocaine-app/infrastructure.py

```diff
@@ -167,158 +166,244 @@ def _sync_state(self):
+                        ):
+                            break
+                    else:
+                        logger.info(
+                            'Removing {} from group {} due to manual group detaching'.format(
+                                nb, group.group_id
+                            )
+                        )
+                        group.remove_node_backend(nb)
+                        group.update_status_recursive()
+        self._sync_ts = new_ts
+
+    def update_group_history(self, group):
+        """
+        Add group to update group history queue
+
+        This method should be called if there is a chance that group has changed
+        it's state and it's group history record needs to be updated.
+        This can happen when:
+            - node backend is added to group;
+            - existing node backend changes its path;
+            - group is added to couple.
+        """
+        with self._groups_to_update_lock:
+            self._groups_to_update.add(group)
+
+    def _new_group_history(self, group_id):
+        gh = GroupHistory(group_id=group_id)
+        gh.collection = self.group_history_finder.collection
+        return gh
+
+    def _new_group_node_backends_set_record(self, group, group_history):
+        """
+        Create new node backends set record for group history
+
+        New record is not allowed to exclude any of the backends that are recorded
+        in the history's most recent set. The sole purpose of new record generation
+        is to extend mentioned set with a new backend that wasn't already there but
+        should be according to the current group's set.
+
+        The workflow is following:
+
+            1) if group's current state contains node backends that are not found
+            in the most recent history record, a new record should be created containing
+            all node backends from the most recent history record as well as node backends
+            from the current group's state;
+
+            2) if group's current state contains no node backends, new record should not
+            be created (this can happen if the group is temporarily unavailable).
+
+        Returns:
+            - new node backends set record to save;
+            - <None> if node backends set should not be updated.
+        """
+
+        new_node_backends_set = None
+
+        storage_nodes = GroupNodeBackendsSet(
+            GroupNodeBackendRecord(**{
+                'hostname': nb.node.host.hostname,
+                'port': nb.node.port,
+                'family': nb.node.family,
+                'backend_id': nb.backend_id,
+                'path': nb.base_path,
+            })
+            for nb in group.node_backends if nb.stat
+        )
+
+        if not storage_nodes:
+            return None
+
+        history_nodes = (
+            group_history.nodes and group_history.nodes[-1] or
+            GroupNodeBackendsSetRecord(set=GroupNodeBackendsSet())
+        )
```

Comment:
What about
```
if group_history.nodes:
    last_history_node = group_history.nodes[-1]
else:
    last_history_node = GroupNodeBackendsSetRecord(set=GroupNodeBackendsSet())
```

`history_nodes` looks like a bad name for this variable. What is it actually about?